Added SASL connect guide for Emacs ERC (#140)
Libera-Chat/libera-chat.github.io: content/_guides/sasl/emacs-erc.md

---
title: Configuring SASL for Emacs ERC
category: sasl

> credits: androclus

Author: okay, i've removed this

---

[ERC](https://www.emacswiki.org/emacs/ERC) is an IRC client which comes as a package for Emacs.

> ## A. Set up ERC IRC client

Reviewer: please avoid numbered headings

Author: okay, these are removed. not sure entirely how to implement anchors and cross-references in github's markdown though (see below) but I did my best (at this point).

> If you have ERC already installed and running on your emacs system, then skip to part B.

Reviewer: Without numbered headings you could just link to them instead

```suggestion
If you have ERC already installed and running on your emacs system, then skip to
[registration](#registration).
```

Author: Okay, I've redone these. I am assuming that being on github.io, these ## sections will get automatically anchored to #user-content-*, as toto_tico's comment at https://stackoverflow.com/questions/5319754/cross-reference-named-anchor-in-markdown seems to indicate. Otherwise, let me know another way you would like me to mark down the cross-reference anchor.

Reviewer: I believe the `user-content-` prefix only applies to content displayed on github.com, but for the site it's processed with Jekyll it doesn't get any prefix, the links you added looks correct if the prefix is removed

Author: Okay, these removed

Otherwise, set up your ERC with basic functionality, as follows:

> 1. Download and Install the erc package from melpa, etc. (*M-x package-list-packages*)

Reviewer: Should this perhaps be escaped as code/commands rather than emphasised?
```suggestion
1. Download and Install the erc package from melpa, etc. (`M-x package-list-packages`)
```

Author: okay, changed as you recommend

2. Set up your ERC for plain (unregistered) sign-in.
  - There are some basic setup instructions [here](https://www.emacswiki.org/emacs/ERC)

>   - Below is my own setup (before adding the SASL stuff below), though by no means do you need all this to set up a basic ERC for yourself.
> ```

Reviewer: Add a blank line before code-block starts, als mark all code blocks with a language. `emacs` or `elisp` is aliases I believe

````suggestion
  - Below is my own setup (before adding the SASL stuff below), though by no means do you need all this to set up a basic ERC for yourself.

```emacs
````

Author: Okay, this is done

;; Most of these are just for convenience options. I believe the only
;; absolutely necessary one here is the my-erc-connect() function I
;; define for connecting.

;; for auto-joining networks and channels when starting up ERC
(require 'erc-join)
(erc-autojoin-enable)

;; Logging of conversations (mostly for when I forget to save peoples’
;; answers when I ask a programming question).

;; turn on logging
(setq erc-log-mode t)
;; write to log file when new text is added to a logged ERC buffer.
(setq erc-log-write-after-insert t)
;; write to log file after every message I send.
(setq erc-log-write-after-send t)
;; Include network name in the buffers
(setq erc-rename-buffers t)

;; from local erc-log.el and tweeked
;; see also https://www.emacswiki.org/emacs/ErcLogging
(require 'erc-log)
(files--ensure-directory "~/.config")
(files--ensure-directory "~/.config/erc")
(files--ensure-directory "~/.config/erc/logs")
(setq erc-log-channels-directory "~/.config/erc/logs/")
(erc-log-enable)

;; don't ask me for my nickserv password
(setq erc-prompt-for-nickserv-password nil)

;; A list of channels to auto-join
(setq erc-autojoin-channels-alist

> 	'(("irc.libera.chat" "#emacs" "#guile" "#scheme" "#erc" "#godot")))

Reviewer: Maybe use an example set of channels related to ERC and Libera?

```suggestion
	'(("irc.libera.chat" "#libera" "#emacs" "#erc")))
```

Author: okay, changed as you recommend

;; OR, if NONE..
;; (setq erc-autojoin-channels-alist nil)

;; from http://www.john2x.com/emacs.html
;; Set the prompt to use the channel name.
(setq erc-prompt  (lambda () (concat (buffer-name) "> ")))

;; beautify the text on screen
(setq erc-fill-function 'erc-fill-static)
(setq erc-fill-static-center 10)

;; my main function for connecting to libera.chat
;; (see also https://libera.chat/guides/connect for regional hostnames)
(defun my-erc-connect ()
  (interactive)

>   (erc :server "irc.libera.chat" :port 6667 :nick "MYNICK" :full-name "MY NAME"))

Reviewer 2: We shouldn't recommend to people to connect to port `6667`. Does ERC support TLS?

Author: Yes, ERC supports TLS. And I can change the port to 6697, will that work?

I can make these two changes, and resubmit with a new pull request -- will that work for you?

(global-set-key (kbd "C-c e") 'my-erc-connect)

;; Channel-tracking: Convenience functions for easily flipping between channels you are logged into
;; from https://www.emacswiki.org/emacs/ErcChannelTracking#toc8
;; (I don't use so much anymore, actually..)
(defvar erc-channels-to-visit nil
    "Channels that have not yet been visited by erc-next-channel-buffer")
(defun erc-next-channel-buffer ()
  "Switch to the next unvisited channel. See erc-channels-to-visit"
  (interactive)
  (when (null erc-channels-to-visit)
    (setq erc-channels-to-visit 
          (remove (current-buffer) (erc-channel-list nil))))
  (let ((target (pop erc-channels-to-visit)))
    (if target 
        (switch-to-buffer target))))
(global-set-key (kbd "C-c n") 'erc-next-channel-buffer)
;; ======== ERC END ================================

> ```
> 3. In the my-erc-connect command you've defined, replace MYNICK with your nickname you plan to use (assuming no one else uses and has registered this nick).

Reviewer: leave a blank row between the end of the code fence and the next row
````suggestion
```

3. In the my-erc-connect command you've defined, replace MYNICK with your nickname you plan to use (assuming no one else uses and has registered this nick).
````

Author: okay, done

> 3. In the my-erc-connect command you've defined, replace MYNICK with your nickname you plan to use (assuming no one else uses and has registered this nick).

Reviewer: I'd escape this
```suggestion
3. In the my-erc-connect command you've defined, replace `MYNICK` with your nickname you plan to use (assuming no one else uses and has registered this nick).
```

Author: okay, done

4. Either shut down your emacs and restart, or evaluate all the above lines.

> ## B. Register

Reviewer: I don't think this section should be included, I'd rather just link to `/guides/registration` in the first part and not include the other steps - this is so we keep the maintenance surface slim in the repo and don't forget to update here if something changes. It's also inconsistent with all other sasl guides right now

Author: I have restructured the page, and made this section a one-liner.

But this actually opens up a larger issue with the guides pages. IMHO, the overall structure of the docs is unclear for newbies.

The problem is that to set up things with SASL on LC, new users need to go through a 6-step process (as below), and that process is not clearly listed currently:

a) Set up your client for connection to libera.chat
b) Log in to libera.chat for the first time, unregistered and over a non-VPN connection
c) Use this connection opportunity to register a permanent nick and password (/_guides/registration.md)
d) Disconnect
e) Reconfigure user's client both i) resetting it for SASL login and ii) entering the new credentials
f) Reconnect and Test, logging in with new SASL client config and credentials

The individual /_guides/sasl/*.md docs pages should all make it clear that these are the 6 main steps, and then organize their sections on each page along these 6 -- just as I am doing.

For instance, the individual /_guides/sasl/*.md doc pages could have the responsibility of laying out steps a, b, d, e, and f. And for step c they could all just link to the generic /guides/registration page (as you rightly suggest, and as I now do).

But also, the /_guides/registration page should make it clear that it is *only one part* -- part "c" -- of a 6-step process, and refer users back to the sasl list page (/_guides/sasl.md).

Also, I would not agree that my page here for Emacs ERC is inconsistent with all other sasl guides right now. The offline Konversation client page (at http://userbase.kde.org/Konversation/Configuring_SASL_authentication) has a whole paragraph in parentheses in Step 9 (starting with "A short interlude:", and containing, "this guide assumes that you already have a Libera Chat account"..). Yes, this page is on a separate server, but the point is still the same: That the author still felt the need to fill in the assumption (that one has to do steps "a" and "b" above).

Since it's a bit of a maze for a new user to do all these steps, I really do think all the docs should really be re-worked and this 6-step process made explicit on every guide page, and with this ERC sasl guide I am providing a model for how to do it.

If you have registered a nickname with libera already, and you remember the password, then skip to part C.
Otherwise, register a nick with libera.chat, as follows:

1. NOTE: In order to connect to irc.libera.chat without SASL (as you'll need to do here in order to register in the first place, right?) you can't be using a VPN. See [our notes on IP Range restrictions](guides/sasl#sasl-access-only-ip-ranges). Later, once you connect via SASL, below, you'll be able to use your VPN to connect.
2. Connect to irc.libera.chat (unregistered), using the nick you plan to register. (You should be able to use the global key C-c n using the global-set-key you defined above to connect).
3. Make sure the nick you plan to use is allowed. (libera's NickServ will give you an error if you are trying to use a nick that is already registered.)
4. Register: Follow our instructions here to [register a nickname and password](/guides/registration) while signed in.
5. Once your registration appears successful, then save your username/password to your password safe of preference (Unix *pass*, or *keepassxc*, etc.).

> ## C. Install ERC-SASL functionality

Reviewer: There should be an empty line after headings

```suggestion
## C. Install ERC-SASL functionality

```

Author: i can't tell the difference here .. (?) sorry

1. Download a copy of erc-sasl.el from [Sylvain Benner's GitHub Repo](https://github.com/syl20bnr/spacemacs/blob/master/layers/%2Bchat/erc/local/erc-sasl/erc-sasl.el). (Note-1)
2. Save it into one of the directorys listed in your emacs’ *load-path* variable. (Note-2)

> 3. Add the following to your .emacs or init.el file in your ERC section (Note-3).
> ```

Reviewer: blank line between, add language

````suggestion
3. Add the following to your .emacs or init.el file in your ERC section (Note-3).

```emacs
````

Author: done

;; Require ERC-SASL package
(require 'erc-sasl)

;; Add SASL server to list of SASL servers (start a new list, if it did not exist)
(add-to-list 'erc-sasl-server-regexp-list "irc\\.libera\\.chat")

;; Redefine/Override the erc-login() function from the erc package, so that
;; it now uses SASL
(defun erc-login ()
  "Perform user authentication at the IRC server. (PATCHED)"
  (erc-log (format "login: nick: %s, user: %s %s %s :%s"
           (erc-current-nick)
           (user-login-name)
           (or erc-system-name (system-name))
           erc-session-server
           erc-session-user-full-name))
  (if erc-session-password
      (erc-server-send (format "PASS %s" erc-session-password))
    (message "Logging in without password"))
  (when (and (featurep 'erc-sasl) (erc-sasl-use-sasl-p))
    (erc-server-send "CAP REQ :sasl"))
  (erc-server-send (format "NICK %s" (erc-current-nick)))
  (erc-server-send
   (format "USER %s %s %s :%s"
       ;; hacked - S.B.
       (if erc-anonymous-login erc-email-userid (user-login-name))
       "0" "*"
       erc-session-user-full-name))
  (erc-update-mode-line))

> ```
> 4. Change the already-existing connection function you had above, so that it now includes the password you registered

Reviewer: ````suggestion
```

4. Change the already-existing connection function you had above, so that it now includes the password you registered
````

Author: done

> 4. Change the already-existing connection function you had above, so that it now includes the password you registered
> ```

Reviewer: ````suggestion
4. Change the already-existing connection function you had above, so that it now includes the password you registered

```emacs
````

Author: done

(defun my-erc-connect ()
  (interactive)
  (erc-tls :server "irc.libera.chat" :port 6697 :nick "NICKNAME" :full-name "MY FULL NAME" :password "PASSWORD"))
(global-set-key (kbd "C-c e") 'my-erc-connect)
```

> ## D. Test

Reviewer: empty line before and after headings
```suggestion

## D. Test

```

Author: okay, changed as you recommend

> 1. Either close down your emacs completely and restart, or evaluate the lines you added, above, including re-evaluating the my-erc-connect function, and try logging in via SASL, again via your *my-erc-connect* shortcut.

Reviewer: commands etc. should be escaped with backticks
```suggestion
1. Either close down your emacs completely and restart, or evaluate the lines you added, above, including re-evaluating the my-erc-connect function, and try logging in via SASL, again via your `my-erc-connect` shortcut.
```

Author: okay, changed as you recommend

> ## E. Notes

Reviewer: ```suggestion
## E. Notes

```

Author: i do not understand this one -- it looks like no change (?)

> - Note-1: If Sylvain’s repo happens to be deleted by the time you read this, then a back-up location is [psachin's gitlab repo](https://gitlab.com/psachin/erc-sasl). The erc-sasl.el there is exactly the same, with one exception: You'll need to change line 54's "loop" to read "cl-loop" (or macro/alias the former to the latter, as some folks do, because cl-loop is a Common Lisp function):

Reviewer: maybe it's better if we just rehost the file if theres a risk of it dissapearing. we already do so with the required chatzilla plugin, see `/static/files`. Alternatively the repo could perhaps be submitted to [Software Heritage](https://archive.softwareheritage.org/save/) as an archive location

Author: I will leave that up to you. But it seems that while it does have the upside of preserving the file, on the other hand, a) it might have copyright issues, and b) if the file is improved upon at its normal github locations, it won't be reflected in the libera copy here, aka a duplication of effort. Still, up to you. I wouldn't want to have to worry about it though, myself.

```
   (cl-loop for re in erc-sasl-server-regexp-list
```

> - Note-2: In my case, I saved it to ~/jeff/.emacs.d/elisp/erc-sasl.el because ~/jeff/.emacs.d/elisp/ directory is on my emacs’ *load-path*.

Reviewer: is Load-path a variable/config setting in emacs? if so it should use backticks rather than **

Author: (Yes, it is). Okay, changed, as you recommend.

;; (I don't use so much anymore, actually..)
(defvar erc-channels-to-visit nil
    "Channels that have not yet been visited by erc-next-channel-buffer")
(defun erc-next-channel-buffer ()
  "Switch to the next unvisited channel. See erc-channels-to-visit"
  (interactive)
  (when (null erc-channels-to-visit)
    (setq erc-channels-to-visit 
          (remove (current-buffer) (erc-channel-list nil))))
  (let ((target (pop erc-channels-to-visit)))
    (if target 
        (switch-to-buffer target))))
(global-set-key (kbd "C-c n") 'erc-next-channel-buffer)
;; ======== ERC END ================================
```

3. In the `my-erc-connect` command you've defined, replace `MYNICK` with your nickname you plan to use (assuming no one else uses and has registered this nick).
4. Either shut down your emacs and restart, or evaluate all the above lines.

## Connect for the First Time

If you have registered a nickname with libera already, and you remember the password, then skip to [Reconfigure
your ERC for SASL functionality](#user-content-reconfigure-your-erc-for-sasl-functionality).

Otherwise, connect for the first time (unregistered), as follows:

> 1. First, make sure you are not using a VPN to connect to libera.chat. In most cases, you won't be allowed to connect unregistered and "un-SASL'd", from a VPN. (See [our notes on IP Range restrictions](guides/sasl#sasl-access-only-ip-ranges).) Later, once you connect as a registered user via SASL, below, you'll be able to use your VPN to connect.

Reviewer: ```suggestion
1. First, make sure you are not using a VPN to connect to libera.chat. In most cases, you won't be allowed to connect unregistered and "un-SASL'd", from a VPN. (See [our notes on IP Range restrictions](/guides/sasl#sasl-access-only-ip-ranges).) Later, once you connect as a registered user via SASL, below, you'll be able to use your VPN to connect.
```

This link should be absolute

Author: fixed.
